homeassistant/components/ads/__init__.py

```diff
@@ -210,3 +216,75 @@ def _device_notification_callback(self, notification, name):
             _LOGGER.warning("No callback available for this datatype")
 
         notification_item.callback(notification_item.name, value)
+
+
+class AdsEntity(Entity):
+    """Representation of ADS entity."""
+
+    def __init__(self, ads_hub, name, ads_var):
+        """Initialize ADS binary sensor."""
+        self._name = name
+        self._unique_id = ads_var
+        self._state_dict = dict()
+        self._state_dict[STATE_KEY_STATE] = None
+        self._ads_hub = ads_hub
+        self._ads_var = ads_var
+        self._event = None
+
+    async def async_initialize_device(self, ads_var, plctype,
+                                      state_key=STATE_KEY_STATE,
+                                      factor=None):
+        """Register device notification."""
+
+        def update(name, value):
+            """Handle device notifications."""
+            _LOGGER.debug('Variable %s changed its value to %d', name, value)
+
+            if factor is None:
+                self._state_dict[state_key] = value
+            else:
+                self._state_dict[state_key] = value / factor
+
+            asyncio.run_coroutine_threadsafe(async_event_set(), self.hass.loop)
+            self.schedule_update_ha_state()
+
+        async def async_event_set():
+            """Set event in async context."""
+            self._event.set()
+
+        self._event = asyncio.Event()
+
+        await self.hass.async_add_executor_job(
+            self._ads_hub.add_device_notification,
+            ads_var, plctype, update)
+        try:
+            with async_timeout.timeout(10):
+                await self._event.wait()
+        except asyncio.TimeoutError:
+            _LOGGER.debug('Variable %s: Timeout during first update',
+                          ads_var)
+
+    @property
+    def name(self):
+        """Return the default name of the binary sensor."""
+        return self._name
+
+    @property
+    def unique_id(self):
+        """Return an unique identifier for this entity."""
+        return self._unique_id
+
+    @property
+    def is_on(self):
```

Comment:
I don't think we should put this property in the common parent class since not all children support it.